File: sdk/typescript/src/stream.ts
import type {
  ConnectionClosingEvent,
  ConversationMessage,
  MessageDeltaEvent,
  StreamResyncEvent,
  TranscriptUpdateEvent,
  TurnChange,
} from "./generated/models/index.js";
import {
  ConnectionClosingEventFromJSON,
  instanceOfConnectionClosingEvent,
  instanceOfConversationMessage,
  instanceOfMessageDeltaEvent,
  instanceOfStreamResyncEvent,
  instanceOfTranscriptUpdateEvent,
  instanceOfTurnChange,
  MessageDeltaEventFromJSON,
  StreamResyncEventFromJSON,
  TranscriptUpdateEventFromJSON,
} from "./generated/models/index.js";
import { ResponseError } from "./generated/runtime.js";
import { isTurnOver } from "./turn-status.js";
import { NvokenError } from "./turn-error.js";

export interface StreamMetadata {
  sseId?: string;
  retryMs?: number;
}

type TypedTurnChange<TOutput extends object> = Omit<TurnChange, "structuredOutput"> & {
  structuredOutput: TOutput | null;
};

type TypedTranscriptUpdate<TOutput extends object> = Omit<TranscriptUpdateEvent, "turnChanges"> & {
  turnChanges: TypedTurnChange<TOutput>[];
};

export type StreamFrame<TOutput extends object = Record<string, never>> = StreamMetadata & (
  | TypedTranscriptUpdate<TOutput>
  | MessageDeltaEvent
  | StreamResyncEvent
  | ConnectionClosingEvent
);

export interface StreamPreview {
  turnId: string;
  attempt: number;
  messageId: string;
  contentIndex: number;
  kind: string;
  delta: string;
  toolCallId?: string;
  name?: string;
}

export interface ReducedSnapshot<TOutput extends object = Record<string, never>> {
  messages: ConversationMessage[];
  turnChanges: TypedTurnChange<TOutput>[];
  previews: StreamPreview[];
  cursor?: string;
}

export interface ReducerOptions<TOutput extends object = Record<string, never>> {
  /** Seed a bounded tail and the exact forward resume position it observed. */
  initial?: Partial<ReducedSnapshot<TOutput>>;
  /** Bound durable messages. Eviction removes whole terminal Turns only. */
  maxMessages?: number;
  /** Bound how many live content previews are retained at once. */
  maxPreviews?: number;
  /** Bound each accumulated preview in UTF-8 bytes. */
  maxPreviewBytes?: number;
}

/**
 * Folds durable frames and provisional previews into one current view.
 *
 * A Turn's lifecycle is folded, not logged: the reducer keeps the highest
 * revision it has seen for each Turn and nothing earlier. Every consumer in
 * this repository already reduced the log that way before publishing it, and
 * the `current` flag cannot substitute for the fold — it means "current as of
 * the read that produced this frame", so two frames for one Turn leave two
 * entries both claiming to be current.
 */
export class Reducer<TOutput extends object = Record<string, never>> {
  private readonly messages = new Map<number, ConversationMessage>();
  private readonly changes = new Map<string, TypedTurnChange<TOutput>>();
  private readonly previews = new Map<string, StreamPreview>();
  private readonly latestAttempts = new Map<string, number>();
  private readonly terminalTurns = new Set<string>();
  private cursor?: string;
  private readonly maxMessages?: number;
  private readonly maxPreviews?: number;
  private readonly maxPreviewBytes?: number;

  constructor(options: ReducerOptions<TOutput> = {}) {
    this.maxMessages = positiveBound(options.maxMessages, "maxMessages");
    this.maxPreviews = positiveBound(options.maxPreviews, "maxPreviews");
    this.maxPreviewBytes = positiveBound(options.maxPreviewBytes, "maxPreviewBytes");
    const initial = options.initial;
    if (!initial) return;
    for (const message of initial.messages ?? []) {
      this.messages.set(message.sequence, message);
    }
    for (const change of initial.turnChanges ?? []) this.storeChange(change);
    for (const preview of initial.previews ?? []) {
      this.previews.set(`${preview.messageId}:${preview.contentIndex}`, { ...preview });
    }
    this.cursor = initial.cursor;
    this.enforceBounds();
  }

  /**
   * Folds a REST transcript page in without moving the stream cursor.
   *
   * This is how a live consumer prepends older history: the page is older than
   * everything the stream has delivered, so adopting its position would replay
   * the conversation from the middle.
   *
   * Bounds are not enforced here. Eviction is oldest-first, and the page being
   * merged is the oldest thing in the window, so enforcing them would discard
   * exactly what was just fetched. A caller with a bound checks the window has
   * room before merging, as the conversation controller does.
   */
  merge(snapshot: Pick<ReducedSnapshot<TOutput>, "messages" | "turnChanges">): void {
    for (const message of snapshot.messages) this.messages.set(message.sequence, message);
    for (const change of snapshot.turnChanges) this.storeChange(change);
  }

  apply(frame: StreamFrame<TOutput>): void {
    if (frame.type === "message.delta") {
      this.appendPreview(frame);
      return;
    }
    if (frame.type === "stream.resync") {
      if (frame.turnId) this.discardPreviews(frame.turnId);
      else {
        this.previews.clear();
        this.latestAttempts.clear();
      }
      return;
    }
    if (frame.type !== "transcript.update") return;

    // A saved message replaces only the preview that was building it. The
    // Turn's next message may already be accumulating, and dropping that
    // prefix loses text no later delta restores.
    for (const message of frame.messages) {
      this.messages.set(message.sequence, message);
      this.discardMessagePreviews(message.id);
    }
    for (const change of frame.turnChanges) {
      this.storeChange(change);
      if (isTurnOver(change)) this.discardPreviews(change.turnId);
    }
    this.cursor = frame.sseId || frame.cursor;
    this.enforceBounds();
  }

  settled(turnId: string): boolean {
    return this.terminalTurns.has(turnId);
  }

  snapshot(): ReducedSnapshot<TOutput> {
    return {
      messages: [...this.messages.values()].sort((left, right) => left.sequence - right.sequence),
      turnChanges: [...this.changes.values()]
        .sort((left, right) => left.turnId.localeCompare(right.turnId)),
      previews: [...this.previews.values()]
        .map((preview) => ({ ...preview }))
        .sort((left, right) => left.messageId.localeCompare(right.messageId)
          || left.contentIndex - right.contentIndex),
      cursor: this.cursor,
    };
  }

  private storeChange(change: TypedTurnChange<TOutput>): void {
    const current = this.changes.get(change.turnId);
    if (!current || change.revision > current.revision) {
      this.changes.set(change.turnId, change);
    }
    if (isTurnOver(change)) this.terminalTurns.add(change.turnId);
  }

  private appendPreview(delta: MessageDeltaEvent): void {
    if (this.terminalTurns.has(delta.turnId)) return;
    const latestAttempt = this.latestAttempts.get(delta.turnId);
    if (latestAttempt !== undefined && delta.attempt < latestAttempt) return;
    if (latestAttempt === undefined || delta.attempt > latestAttempt) {
      this.discardPreviews(delta.turnId);
      this.latestAttempts.set(delta.turnId, delta.attempt);
    }
    const key = `${delta.messageId}:${delta.contentIndex}`;
    const preview = this.previews.get(key) ?? {
      turnId: delta.turnId,
      attempt: delta.attempt,
      messageId: delta.messageId,
      contentIndex: delta.contentIndex,
      kind: delta.kind,
      delta: "",
    };
    preview.attempt = delta.attempt;
    preview.kind = delta.kind;
    preview.delta = appendUTF8(preview.delta, delta.delta, this.maxPreviewBytes);
    if (delta.toolCallId) preview.toolCallId = delta.toolCallId;
    if (delta.name) preview.name = delta.name;
    this.previews.set(key, preview);
    if (this.maxPreviews === undefined) return;
    // Map iteration is insertion order, so the first key is the oldest
    // preview still accumulating.
    while (this.previews.size > this.maxPreviews) {
      const oldest = this.previews.keys().next().value;
      if (oldest === undefined) break;
      this.previews.delete(oldest);
    }
  }

  private discardPreviews(turnId: string): void {
    for (const [key, preview] of this.previews) {
      if (preview.turnId === turnId) this.previews.delete(key);
    }
    this.latestAttempts.delete(turnId);
  }

  private discardMessagePreviews(messageId: string): void {
    for (const [key, preview] of this.previews) {
      if (preview.messageId === messageId) this.previews.delete(key);
    }
  }

  /**
   * Evicts oldest-first, and only at a Turn boundary that has settled.
   *
   * Dropping half a Turn would leave a transcript that reads as though the
   * agent answered a question nobody asked. A live Turn is never evicted at
   * all: it is still producing the messages that finish it, so the window
   * stops growing down rather than cutting into work in progress.
   */
  private enforceBounds(): void {
    if (this.maxMessages === undefined) return;
    while (this.messages.size > this.maxMessages) {
      const ordered = [...this.messages.values()]
        .sort((left, right) => left.sequence - right.sequence);
      const oldest = ordered[0];
      if (!oldest) return;
      const boundary = oldest.turnId;
      if (!boundary) {
        this.messages.delete(oldest.sequence);
        continue;
      }
      if (!this.terminalTurns.has(boundary)) return;
      for (const message of ordered) {
        if (message.turnId === boundary) this.messages.delete(message.sequence);
      }
      this.changes.delete(boundary);
      this.terminalTurns.delete(boundary);
    }
  }
}

function positiveBound(value: number | undefined, name: string): number | undefined {
  if (value === undefined) return undefined;
  if (!Number.isSafeInteger(value) || value < 1) {
    throw new NvokenError("validation", `${name} must be a positive integer`);
  }
  return value;
}

/**
 * Appends within a UTF-8 byte budget, truncating at a code-point boundary.
 *
 * A preview is text a page renders. Cutting one mid code point produces a
 * replacement character that stays on screen until the saved message replaces
 * the whole preview, which is exactly the kind of glitch a bound meant to
 * prevent problems introduces instead.
 */
function appendUTF8(current: string, added: string, maximum?: number): string {
  if (maximum === undefined) return current + added;
  const combined = current + added;
  const bytes = new TextEncoder().encode(combined);
  if (bytes.length <= maximum) return combined;
  const decoder = new TextDecoder("utf-8", { fatal: true });
  for (let end = maximum; end > 0; end -= 1) {
    try {
      return decoder.decode(bytes.slice(0, end));
    } catch {
      // Step back to the prior code-point boundary.
    }
  }
  return "";
}

export interface StreamLoopOptions {
  cursor?: string;
  signal?: AbortSignal;
  timeoutMs?: number;
  reconnectTimeoutMs?: number;
  /**
   * Reports when the transport is up and when it is about to reconnect.
   *
   * A deliberate close is already visible as a `connection.closing` frame. A
   * silent drop is not, and "reconnecting" is the state a page needs in order
   * to stop looking broken while it heals. Observational only: a throw here
   * cannot change what the stream delivers, so it is swallowed.
   */
  onConnectionChange?: (state: "connected" | "reconnecting") => void;
}

export type StreamConnector = (
  cursor: string | undefined,
  signal: AbortSignal | undefined,
) => Promise<Response>;

export function streamTurnFrames<TOutput extends object = Record<string, never>>(
  connect: StreamConnector,
  options: StreamLoopOptions = {},
): AsyncGenerator<StreamFrame<TOutput>> {
  return reconnectingFrames(connect, options, true);
}

export function streamConversationFrames<TOutput extends object = Record<string, never>>(
  connect: StreamConnector,
  options: StreamLoopOptions = {},
): AsyncGenerator<StreamFrame<TOutput>> {
  return reconnectingFrames(connect, options, false);
}

async function* reconnectingFrames<TOutput extends object>(
  connect: StreamConnector,
  options: StreamLoopOptions,
  stopAtTerminalTurn: boolean,
): AsyncGenerator<StreamFrame<TOutput>> {
  const scope = abortScope(options.signal, options.timeoutMs);
  const reconnectTimeoutMs = options.reconnectTimeoutMs ?? 300_000;
  if (!Number.isFinite(reconnectTimeoutMs) || reconnectTimeoutMs <= 0) {
    scope.dispose();
    throw new TypeError("reconnectTimeoutMs must be a positive finite number");
  }
  let cursor = options.cursor;
  let delayMs = 100;
  let failureStartedAt: number | undefined;
  let serverRetryMs: number | undefined;
  try {
    while (!scope.signal?.aborted) {
      try {
        const response = await connect(cursor, scope.signal);
        notifyConnection(options, "connected");
        failureStartedAt = undefined;
        let requestedReconnect = false;
        for await (const event of parseEventStream(response, scope.signal)) {
          if (event.retryMs !== undefined && Number.isFinite(event.retryMs) && event.retryMs >= 0) {
            serverRetryMs = Math.min(10_000, event.retryMs);
          }
          // A frame containing only `retry:` is SSE control data, not an
          // nvoken event. Apply its reconnect delay without trying to decode it.
          if (event.data === undefined) continue;
          const frame = decodeStreamFrame<TOutput>(event);
          // Added frame types must not break an older reader. Unknown frames
          // carry no state this version knows how to reduce, so skip them.
          if (frame === undefined) continue;
          if (frame.sseId) cursor = frame.sseId;
          if (frame.type === "transcript.update") cursor = frame.cursor || cursor;
          yield frame;
          if (stopAtTerminalTurn
            && frame.type === "transcript.update"
            && frame.turnChanges.some((change) => isTurnOver(change))) return;
          if (frame.type === "connection.closing") {
            requestedReconnect = true;
            break;
          }
        }
        delayMs = serverRetryMs ?? (requestedReconnect ? 0 : 100);
      } catch (error) {
        if (scope.signal?.aborted) throw abortReason(scope.signal);
        if (!isRetryableStreamError(error)) throw error;
        failureStartedAt ??= Date.now();
        if (Date.now() - failureStartedAt >= reconnectTimeoutMs) throw error;
      }
      notifyConnection(options, "reconnecting");
      if (delayMs > 0) await pause(delayMs, scope.signal);
      delayMs = serverRetryMs ?? Math.min(2_000, Math.max(100, delayMs * 2));
    }
    throw abortReason(scope.signal);
  } finally {
    scope.dispose();
  }
}

function notifyConnection(
  options: StreamLoopOptions,
  state: "connected" | "reconnecting",
): void {
  try {
    options.onConnectionChange?.(state);
  } catch {
    // Transport correctness cannot depend on an observational callback.
  }
}

interface ParsedEvent {
  id?: string;
  type: string;
  data?: unknown;
  retryMs?: number;
}

async function* parseEventStream(
  response: Response,
  signal?: AbortSignal,
): AsyncGenerator<ParsedEvent> {
  if (!response.body) throw new TypeError("nvoken stream response has no body");
  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  try {
    while (true) {
      if (signal?.aborted) throw abortReason(signal);
      const { done, value } = await reader.read();
      buffer += decoder.decode(value, { stream: !done });
      let boundary = eventBoundary(buffer);
      while (boundary !== undefined) {
        const raw = buffer.slice(0, boundary.index);
        buffer = buffer.slice(boundary.index + boundary.length);
        const event = parseEvent(raw);
        if (event) yield event;
        boundary = eventBoundary(buffer);
      }
      if (done) {
        const event = parseEvent(buffer);
        if (event) yield event;
        return;
      }
    }
  } finally {
    reader.releaseLock();
  }
}

function eventBoundary(value: string): { index: number; length: number } | undefined {
  const match = /\r?\n\r?\n/.exec(value);
  return match ? { index: match.index, length: match[0].length } : undefined;
}

function parseEvent(raw: string): ParsedEvent | undefined {
  if (!raw.trim()) return undefined;
  const data: string[] = [];
  let id: string | undefined;
  let type = "message";
  let retryMs: number | undefined;
  for (const line of raw.split(/\r?\n/)) {
    if (!line || line.startsWith(":")) continue;
    const separator = line.indexOf(":");
    const field = separator < 0 ? line : line.slice(0, separator);
    const value = separator < 0 ? "" : line.slice(separator + 1).replace(/^ /, "");
    if (field === "data") data.push(value);
    else if (field === "event") type = value;
    else if (field === "id" && !value.includes("\0")) id = value;
    else if (field === "retry" && /^\d+$/.test(value)) retryMs = Number(value);
  }
  const joined = data.join("\n");
  let decoded: unknown;
  if (data.length > 0 && joined) {
    try { decoded = JSON.parse(joined); }
    catch (error) { throw new TypeError(`invalid ${type} stream JSON`, { cause: error }); }
  } else if (data.length > 0) {
    decoded = "";
  }
  return { id, type, data: decoded, retryMs };
}

function decodeStreamFrame<TOutput extends object>(
  event: ParsedEvent,
): StreamFrame<TOutput> | undefined {
  if (!event.data || typeof event.data !== "object") {
    throw new NvokenError(
      "unexpected_response",
      `${event.type} stream frame must contain a JSON object`,
    );
  }
  const metadata = { sseId: event.id, retryMs: event.retryMs };
  switch (event.type) {
    case "transcript.update": {
      const update = requireFrameField(
        "transcript.update",
        TranscriptUpdateEventFromJSON(event.data),
        instanceOfTranscriptUpdateEvent,
      );
      for (const message of update.messages) {
        requireFrameField("conversation message", message, instanceOfConversationMessage);
      }
      for (const change of update.turnChanges) {
        requireFrameField("turn change", change, instanceOfTurnChange);
      }
      return { ...update, ...metadata } as StreamFrame<TOutput>;
    }
    case "message.delta":
      return {
        ...requireFrameField(
          "message.delta",
          MessageDeltaEventFromJSON(event.data),
          instanceOfMessageDeltaEvent,
        ),
        ...metadata,
      } as StreamFrame<TOutput>;
    case "stream.resync":
      return {
        ...requireFrameField(
          "stream.resync",
          StreamResyncEventFromJSON(event.data),
          instanceOfStreamResyncEvent,
        ),
        ...metadata,
      } as StreamFrame<TOutput>;
    case "connection.closing":
      return {
        ...requireFrameField(
          "connection.closing",
          ConnectionClosingEventFromJSON(event.data),
          instanceOfConnectionClosingEvent,
        ),
        ...metadata,
      } as StreamFrame<TOutput>;
    default:
      return undefined;
  }
}

function requireFrameField<T>(
  what: string,
  value: T,
  valid: (candidate: object) => boolean,
): T {
  if (!valid(value as object)) {
    throw new NvokenError(
      "unexpected_response",
      `${what} is missing a field the contract requires`,
    );
  }
  return value;
}

function isRetryableStreamError(error: unknown): boolean {
  if (error instanceof ResponseError) {
    return error.response.status === 408
      || error.response.status === 425
      || error.response.status === 429
      || error.response.status >= 500;
  }
  return error instanceof TypeError;
}

function pause(milliseconds: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted) { reject(abortReason(signal)); return; }
    const onAbort = () => { clearTimeout(timer); reject(abortReason(signal)); };
    const timer = setTimeout(() => {
      signal?.removeEventListener("abort", onAbort);
      resolve();
    }, milliseconds);
    signal?.addEventListener("abort", onAbort, { once: true });
  });
}

interface AbortScope {
  signal?: AbortSignal;
  dispose(): void;
}

function abortScope(signal: AbortSignal | undefined, timeoutMs: number | undefined): AbortScope {
  if (timeoutMs === undefined) return { signal, dispose: () => undefined };
  if (!Number.isFinite(timeoutMs) || timeoutMs <= 0) {
    throw new TypeError("timeoutMs must be a positive finite number");
  }
  const controller = new AbortController();
  const abort = () => controller.abort(signal?.reason);
  if (signal?.aborted) abort();
  else signal?.addEventListener("abort", abort, { once: true });
  const timer = setTimeout(() => controller.abort(new DOMException("Timed out", "AbortError")), timeoutMs);
  return {
    signal: controller.signal,
    dispose: () => { clearTimeout(timer); signal?.removeEventListener("abort", abort); },
  };
}

function abortReason(signal?: AbortSignal): unknown {
  return signal?.reason ?? new DOMException("Aborted", "AbortError");
}

File: sdk/go/stream.go
package nvoken

import (
	"bufio"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/deepnoodle-ai/nvoken/sdk/go/generated"
)

// StreamEvent is one decoded SSE frame. Data contains the exact target event
// JSON, while ID is the durable resume cursor when the frame has one.
type StreamEvent struct {
	ID    string          `json:"id,omitempty"`
	Type  string          `json:"type"`
	Data  json.RawMessage `json:"data"`
	Retry time.Duration   `json:"retry,omitempty"`
}

var ErrStopStream = errors.New("stop reading the stream")

type UpdatesOptions struct {
	Cursor *string
	Deltas *bool
}

type StreamPreview struct {
	TurnID       string `json:"turn_id"`
	Attempt      int64  `json:"attempt"`
	MessageID    string `json:"message_id"`
	ContentIndex int    `json:"content_index"`
	Kind         string `json:"kind"`
	Delta        string `json:"delta"`
	ToolCallID   string `json:"tool_call_id,omitempty"`
	Name         string `json:"name,omitempty"`
}

type ReducedSnapshot struct {
	Messages    []ConversationMessage `json:"messages"`
	TurnChanges []TurnChange          `json:"turn_changes"`
	Previews    []StreamPreview       `json:"previews"`
	Cursor      string                `json:"cursor,omitempty"`
}

// TurnUpdate is the high-level reduced view yielded while following one Turn.
// Previews are provisional deltas; Snapshot.Messages contains durable saved
// messages and Snapshot becomes authoritative after the final point read.
type TurnUpdate struct {
	Snapshot TurnSnapshot
	Previews []StreamPreview
	Cursor   string
}

type streamPreviewKey struct {
	messageID    string
	contentIndex int
}

// Reducer folds replayable durable frames and provisional previews into one
// current view. Durable frames are idempotent by identity and revision.
type Reducer struct {
	messages       map[int64]ConversationMessage
	changes        map[string]TurnChange
	previews       map[streamPreviewKey]StreamPreview
	latestAttempts map[string]int64
	terminalTurns  map[string]struct{}
	cursor         string
}

func NewReducer() *Reducer {
	return &Reducer{
		messages:       make(map[int64]ConversationMessage),
		changes:        make(map[string]TurnChange),
		previews:       make(map[streamPreviewKey]StreamPreview),
		latestAttempts: make(map[string]int64),
		terminalTurns:  make(map[string]struct{}),
	}
}

func (r *Reducer) seed(messages []ConversationMessage) {
	for _, message := range messages {
		r.messages[message.Sequence] = message
	}
}

func (r *Reducer) Apply(event StreamEvent) error {
	switch event.Type {
	case "message.delta":
		if err := requireFrameKeys("MessageDeltaEvent", event.Data); err != nil {
			return err
		}
		var delta generated.MessageDeltaEvent
		if err := json.Unmarshal(event.Data, &delta); err != nil {
			return fmt.Errorf("decode message delta: %w", err)
		}
		r.appendPreview(delta)
		return nil
	case "stream.resync":
		if err := requireFrameKeys("StreamResyncEvent", event.Data); err != nil {
			return err
		}
		var resync generated.StreamResyncEvent
		if err := json.Unmarshal(event.Data, &resync); err != nil {
			return fmt.Errorf("decode stream resync: %w", err)
		}
		if resync.TurnID == nil {
			clear(r.previews)
			clear(r.latestAttempts)
		} else {
			r.discardPreviews(*resync.TurnID)
		}
		return nil
	case "transcript.update":
		if err := requireTranscriptUpdateKeys(event.Data); err != nil {
			return err
		}
		var update generated.TranscriptUpdateEvent
		if err := json.Unmarshal(event.Data, &update); err != nil {
			return fmt.Errorf("decode transcript update: %w", err)
		}
		// Messages land before terminal changes so a snapshot never settles a
		// Turn before its final saved message is visible. A saved message
		// replaces only the preview that was building it: the Turn's next
		// message may already be accumulating, and dropping that prefix loses
		// text no later delta restores.
		for _, message := range update.Messages {
			r.messages[message.Sequence] = message
			r.discardMessagePreviews(string(message.ID))
		}
		for _, change := range update.TurnChanges {
			key := fmt.Sprintf("%s:%d", change.TurnID, change.Revision)
			r.changes[key] = change
			if change.Terminal {
				r.terminalTurns[change.TurnID] = struct{}{}
				r.discardPreviews(change.TurnID)
			}
		}
		if event.ID != "" {
			r.cursor = event.ID
		} else if update.Cursor != "" {
			r.cursor = update.Cursor
		}
	}
	return nil
}

func (r *Reducer) Settled(turnID string) bool {
	_, ok := r.terminalTurns[turnID]
	return ok
}

func (r *Reducer) Snapshot() ReducedSnapshot {
	messages := make([]ConversationMessage, 0, len(r.messages))
	for _, message := range r.messages {
		messages = append(messages, message)
	}
	sort.Slice(messages, func(i, j int) bool { return messages[i].Sequence < messages[j].Sequence })
	changes := make([]TurnChange, 0, len(r.changes))
	for _, change := range r.changes {
		changes = append(changes, change)
	}
	sort.Slice(changes, func(i, j int) bool {
		if changes[i].TurnID == changes[j].TurnID {
			return changes[i].Revision < changes[j].Revision
		}
		return changes[i].TurnID < changes[j].TurnID
	})
	previews := make([]StreamPreview, 0, len(r.previews))
	for _, preview := range r.previews {
		previews = append(previews, preview)
	}
	sort.Slice(previews, func(i, j int) bool {
		if previews[i].MessageID == previews[j].MessageID {
			return previews[i].ContentIndex < previews[j].ContentIndex
		}
		return previews[i].MessageID < previews[j].MessageID
	})
	return ReducedSnapshot{Messages: messages, TurnChanges: changes, Previews: previews, Cursor: r.cursor}
}

func (r *Reducer) appendPreview(delta generated.MessageDeltaEvent) {
	turnID := string(delta.TurnID)
	if _, terminal := r.terminalTurns[turnID]; terminal {
		return
	}
	if latest, ok := r.latestAttempts[turnID]; ok {
		if delta.Attempt < latest {
			return
		}
		if delta.Attempt > latest {
			r.discardPreviews(turnID)
		}
	}
	r.latestAttempts[turnID] = delta.Attempt
	key := streamPreviewKey{messageID: delta.MessageID, contentIndex: delta.ContentIndex}
	preview := r.previews[key]
	preview.TurnID = turnID
	preview.Attempt = delta.Attempt
	preview.MessageID = delta.MessageID
	preview.ContentIndex = delta.ContentIndex
	preview.Kind = string(delta.Kind)
	preview.Delta += delta.Delta
	if delta.ToolCallID != nil {
		preview.ToolCallID = *delta.ToolCallID
	}
	if delta.Name != nil {
		preview.Name = *delta.Name
	}
	r.previews[key] = preview
}

func (r *Reducer) discardPreviews(turnID string) {
	for key, preview := range r.previews {
		if preview.TurnID == turnID {
			delete(r.previews, key)
		}
	}
	delete(r.latestAttempts, turnID)
}

// discardMessagePreviews drops every content index of one previewed message
// once the saved message carrying that ID has landed.
func (r *Reducer) discardMessagePreviews(messageID string) {
	for key := range r.previews {
		if key.messageID == messageID {
			delete(r.previews, key)
		}
	}
}

// Updates follows this Turn as reduced snapshots, reconnecting from the last
// durable cursor until its terminal change arrives or the consumer stops it
// with ErrStopStream. Raw SSE frames remain available through Raw().
func (t *Turn) Updates(ctx context.Context, options UpdatesOptions, consume func(TurnUpdate) error) error {
	if err := t.validateAccess(); err != nil {
		return err
	}
	current, err := t.Status(ctx)
	if err != nil {
		return turnWaitError(err, t)
	}
	if current.Resource.Status == generated.TurnStatusWaiting {
		if _, err := t.settleHostTools(ctx, current.Resource.ToolCalls); err != nil {
			return err
		}
	}
	if consume != nil {
		if err := consume(TurnUpdate{Snapshot: *current}); err != nil {
			if errors.Is(err, ErrStopStream) {
				return nil
			}
			return err
		}
	}
	if current.Resource.EndedAt != nil {
		return nil
	}
	reducer := NewReducer()
	reducer.seed(current.Messages)
	retryDelay := time.Second
	for {
		cursor := options.Cursor
		lastEventID := reducer.Snapshot().Cursor
		params := &generated.StreamTurnParams{Deltas: options.Deltas}
		if lastEventID != "" {
			params.LastEventID = &lastEventID
		} else if cursor != nil {
			value := generated.Cursor(*cursor)
			params.Cursor = &value
		}
		response, err := t.client.raw.ClientInterface.StreamTurn(ctx, t.id, params, t.requestEditor())
		if err != nil {
			if err := waitForReconnect(ctx, retryDelay); err != nil {
				return turnWaitError(transportError(err), t)
			}
			continue
		}
		if response.StatusCode < 200 || response.StatusCode >= 300 {
			body, _ := io.ReadAll(io.LimitReader(response.Body, 1<<20))
			_ = response.Body.Close()
			return errorFromResponse(response.StatusCode, response.Header, body)
		}
		err = readSSE(response.Body, func(event StreamEvent) error {
			if event.Retry > 0 {
				retryDelay = event.Retry
			}
			if len(event.Data) == 0 {
				return nil
			}
			if err := reducer.Apply(event); err != nil {
				return err
			}
			if event.Type == "transcript.update" {
				var update generated.TranscriptUpdateEvent
				if err := json.Unmarshal(event.Data, &update); err != nil {
					return fmt.Errorf("decode transcript update for tools: %w", err)
				}
				for _, change := range update.TurnChanges {
					if change.TurnID != t.id {
						continue
					}
					applyTurnChange(&current.Resource, change)
					if change.Status == generated.TurnStatusWaiting && change.ToolCalls != nil {
						if _, err := t.settleHostTools(ctx, *change.ToolCalls); err != nil {
							return err
						}
					}
				}
			}
			reduced := reducer.Snapshot()
			current.Messages = reduced.Messages
			if consume != nil {
				if err := consume(TurnUpdate{
					Snapshot: *current,
					Previews: reduced.Previews,
					Cursor:   reduced.Cursor,
				}); err != nil {
					return err
				}
			}
			if reducer.Settled(t.id) {
				final, err := t.Status(ctx)
				if err != nil {
					return turnWaitError(err, t)
				}
				if consume != nil {
					if err := consume(TurnUpdate{Snapshot: *final, Cursor: reduced.Cursor}); err != nil {
						return err
					}
				}
				return ErrStopStream
			}
			return nil
		})
		_ = response.Body.Close()
		if errors.Is(err, ErrStopStream) {
			return nil
		}
		if err != nil && err != io.EOF {
			return err
		}
		if err := waitForReconnect(ctx, retryDelay); err != nil {
			return turnWaitError(transportError(err), t)
		}
	}
}

func applyTurnChange(turn *TurnResource, change TurnChange) {
	turn.Status = change.Status
	turn.ConversationID = change.ConversationID
	turn.ContentExpiresAt = change.ContentExpiresAt
	turn.CreditBlock = change.CreditBlock
	turn.Error = change.Error
	turn.Provenance = change.Provenance
	turn.StopReason = change.StopReason
	turn.StructuredOutput = change.StructuredOutput
	turn.StructuredOutputProvenance = change.StructuredOutputProvenance
	turn.Usage = change.Usage
	turn.UpdatedAt = change.OccurredAt
	if change.ToolCalls != nil {
		turn.ToolCalls = append([]ToolCallSummary(nil), (*change.ToolCalls)...)
	}
	if change.Terminal && turn.EndedAt == nil {
		endedAt := change.OccurredAt
		turn.EndedAt = &endedAt
	}
}

func readSSE(reader io.Reader, consume func(StreamEvent) error) error {
	scanner := bufio.NewScanner(reader)
	scanner.Buffer(make([]byte, 4096), 2<<20)
	event := StreamEvent{}
	var data []string
	dispatch := func() error {
		if event.Type == "" && len(data) == 0 && event.ID == "" && event.Retry == 0 {
			return nil
		}
		event.Data = json.RawMessage(strings.Join(data, "\n"))
		if err := consume(event); err != nil {
			return err
		}
		event = StreamEvent{}
		data = nil
		return nil
	}
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			if err := dispatch(); err != nil {
				return err
			}
			continue
		}
		if strings.HasPrefix(line, ":") {
			continue
		}
		field, value, found := strings.Cut(line, ":")
		if found && strings.HasPrefix(value, " ") {
			value = value[1:]
		}
		switch field {
		case "event":
			event.Type = value
		case "id":
			event.ID = value
		case "data":
			data = append(data, value)
		case "retry":
			if milliseconds, err := strconv.Atoi(value); err == nil && milliseconds >= 0 {
				event.Retry = time.Duration(milliseconds) * time.Millisecond
			}
		}
	}
	if err := scanner.Err(); err != nil {
		return err
	}
	if len(data) > 0 || event.Type != "" || event.ID != "" || event.Retry > 0 {
		if err := dispatch(); err != nil {
			return err
		}
	}
	return io.EOF
}

func waitForReconnect(ctx context.Context, delay time.Duration) error {
	if delay <= 0 {
		delay = time.Second
	}
	if delay > 30*time.Second {
		delay = 30 * time.Second
	}
	timer := time.NewTimer(delay)
	select {
	case <-ctx.Done():
		timer.Stop()
		return transportError(ctx.Err())
	case <-timer.C:
		return nil
	}
}
